### ISSUE TITLE
Question:Overriding run_trial() method

### ISSUE THREAD

**Issue by shakib221b** [keras-team/keras-tuner#199]
Sorry if I am asking very trivial question, I am not good at OOP or programming in general. I wanted to add hyperparameters during data augmentation. So I tried to override run_trial() method to do hypertuning(as shown in the tutorial) and was successful. But my code doesn't need any custom training or validation loop. A model.fit() would probably suffice. Is there any way I can write my run_trial() method more efficiently? Here's a code snippet:

`def get_data(hp, X_train, X_valid):

    """
    Gets data from train and valid dataframe
    and performs various augmentations
    """
    train_gen = ImageDataGenerator(
        rotation_range=hp.Int('rotation', 0, 30, step=5, default=10),
        width_shift_range=hp.Float('w_shift', 0.0, 0.4, step=0.1, default=0.0),
        height_shift_range=hp.Float('h_shift', 0.0, 0.4, default=0.0),
        shear_range=hp.Float('shear', 0.0, 30.0, default=0),
        zoom_range=hp.Float('zoom', 0.0, 0.5, default=0.0),
        fill_mode='nearest',
        horizontal_flip=hp.Choice('h_flip', [True, False], default=True),
        rescale=1./255.)

    valid_gen = ImageDataGenerator(rescale=1./255.)

    train_data = train_gen.flow_from_dataframe(
            dataframe=X_train,
            x_col='filename',
            y_col='class',
            class_mode="categorical",
            target_size=(224, 224),
            batch_size=hp.Int('batch_size', 32, 128, step=32, default=64))

    valid_data = valid_gen.flow_from_dataframe(
            dataframe=X_valid,
            x_col='filename',
            y_col='class',
            class_mode="categorical",
            target_size=(224, 224),
            batch_size=32)

    print('Creating datagenerator...')

    return train_data, valid_data

class MyTuner(kt.Tuner):

    def run_trial(self, trial, X_train, X_valid, epochs=10):
        hp = trial.hyperparameters

        # Hyperparameters can be added anywhere inside `run_trial`.
        # When the first trial is run, they will take on their default values.
        # Afterwards, they will be tuned by the `Oracle`.

        cnt = 0
        check = 0
        train_data, valid_data = get_data(hp, X_train, X_valid)
        train_batches = train_data.n//train_data.batch_size
        valid_batches = valid_data.n//valid_data.batch_size

        model = self.hypermodel.build(trial.hyperparameters)

        if hp['optimizer'] == 'adam':
            optimizer = keras.optimizers.Adam(
                learning_rate=hp.Float(
                    'lr_adam', 1e-4, 1e-3, sampling='log', default=1e-3))
        else:
            optimizer = keras.optimizers.SGD(
                learning_rate=hp.Float(
                    'lr_SGD', 0.001, 0.01, sampling='log', default=0.01),
                momentum=hp.Float('lr_SGD_m', 0.0, 0.4, default=0.0))

        train_loss_metric = tf.keras.metrics.Mean()
        valid_loss_metric = tf.keras.metrics.Mean()

        train_acc_metric = keras.metrics.CategoricalAccuracy()
        valid_acc_metric = keras.metrics.CategoricalAccuracy()

        @tf.function
        def run_train_step(images, labels):
            with tf.GradientTape() as tape:
                pred = model(images)
                loss = tf.keras.losses.categorical_crossentropy(
                    labels, pred)
                gradients = tape.gradient(loss, model.trainable_variables)
            optimizer.apply_gradients(
                zip(gradients, model.trainable_variables))

            train_loss_metric.update_state(loss)
            train_acc_metric.update_state(labels, pred)
            return loss

        @tf.function
        def run_valid_step(images, labels):
            pred = model(images)
            loss = tf.keras.losses.categorical_crossentropy(
                labels, pred)

            valid_loss_metric.update_state(loss)
            valid_acc_metric.update_state(labels, pred)
            return loss

        # `self.on_epoch_end` reports results to the `Oracle` and saves the
        # current state of the Model.
        # The other hooks called here only log values for display
        # but can also be overridden.
        # For use cases where there is no natural concept of epoch,
        # you do not have to call any of these hooks.
        # In this case you should instead call `self.oracle.update_trial` and
        # `self.oracle.save_model` manually.

        for epoch in range(10):
            print('Epoch: {} of {}'.format(epoch+1, epochs))

            self.on_epoch_begin(trial, model, epoch, logs={})

            # train Loop
            for batch, (images, labels) in enumerate(train_data):
                self.on_batch_begin(trial, model, batch, logs={})
                batch_loss = run_train_step(images, labels)
                self.on_batch_end(
                    trial, model, batch, logs={'train_loss': batch_loss})

                if batch == train_batches-1:
                    loss = train_loss_metric.result().numpy()
                    acc = train_acc_metric.result().numpy()
                    print(
                        'Train Batches: {}, Avg Loss: {}, Avg Acc: {}'.format(
                            batch, loss, acc))
                    break

            # validation loop
            for batch, (images, labels) in enumerate(valid_data):
                self.on_batch_begin(trial, model, batch, logs={})
                batch_loss = run_valid_step(images, labels)
                self.on_batch_end(
                    trial, model, batch, logs={'valid_loss': batch_loss})

                if batch == valid_batches-1:
                    loss = valid_loss_metric.result().numpy()
                    acc = valid_acc_metric.result().numpy()
                    print(
                        'Valid Batches: {}, Loss: {}, Accuracy: {}'.format(
                            batch, loss, acc))
                    break

            valid_acc = valid_acc_metric.result().numpy()
            if check <= valid_acc:
                check = valid_acc
                cnt = 0
            else:
                cnt += 1
            self.on_epoch_end(
                trial, model, epoch, logs={'valid_acc': valid_acc})

            train_loss_metric.reset_states()
            valid_loss_metric.reset_states()
            train_acc_metric.reset_states()
            valid_acc_metric.reset_states()
            if cnt == 4:
                print("Validation accuracy didn't improve for 5 epochs")
                break
        print("Ending trial")`

`tuner = MyTuner(
        oracle=kt.oracles.BayesianOptimization(
            objective=kt.Objective('valid_acc', 'max'),
            max_trials=mx_trial),
        hypermodel=MyHyperModel(base_model),
        directory='../tuning',
        project_name=base_model)`
`tuner.search(X_train, X_valid, epochs=epochs)
`

**Comment by omalleyt12** [comment]
@shakib221b Thanks for the issue!

Yes I think in this case IIUC, you should be able to perform the data augmentation and then call into `super(MyTuner, self).run_trial`. Something like this should work:


```python
class MyTuner(kt.Tuner):
  def run_trial(self, trial, X_train, X_valid, epochs=10):
    train_data, valid_data = get_data(trial.hyperparameters, X_train, X_valid)
    super(MyTuner, self).run_trial(trial, train_data, validation_data=valid_data, epochs=10)
```

**Comment by omalleyt12** [comment]
Please reopen if that doesn't work

**Comment by shakib221b** [comment]
Thank you so much!
